localstack/utils/scheduler.py

```diff
@@ -0,0 +1,184 @@
+import queue
+import threading
+import time
+from concurrent.futures import Executor
+from typing import Any, Callable, List, Mapping, Optional, Tuple, Union
+
+
+class ScheduledTask:
+    """
+    Internal representation of a task (a callable) and its scheduling parameters.
+    """
+
+    def __init__(
+        self,
+        task: Callable,
+        period: Optional[float] = None,
+        fixed_rate: bool = True,
+        start: Optional[float] = None,
+        on_error: Callable[[Exception], None] = None,
+        args: Optional[Union[tuple, list]] = None,
+        kwargs: Optional[Mapping[str, Any]] = None,
+    ) -> None:
+        super().__init__()
+        self.task = task
+        self.fixed_rate = fixed_rate
+        self.period = period
+        self.start = start
+        self.on_error = on_error
+        self.args = args or tuple()
+        self.kwargs = kwargs or dict()
+
+        self.deadline = None
+        self.error = None
+        self._cancelled = False
+
+    @property
+    def is_periodic(self) -> bool:
+        return self.period is not None
+
+    @property
+    def is_cancelled(self) -> bool:
+        return self._cancelled
+
+    def set_next_deadline(self):
+        if not self.deadline:
+            raise ValueError("Deadline was not initialized")
+
+        if self.fixed_rate:
+            self.deadline = self.deadline + self.period
+        else:
+            self.deadline = time.time() + self.period
+
+    def cancel(self):
+        self._cancelled = True
+
+    def run(self):
+        """
+
+        :return:
+        """
+        try:
+            self.task(*self.args, **self.kwargs)
+        except Exception as e:
+            if self.on_error:
+                self.on_error(e)
+
+
+class Scheduler:
+    """
+    An event-loop based task scheduler that can be parallelized with an executor.
+    """
+
+    POISON = (-1, "__POISON__")
```

Comment:
Love how we're introducing more [shutdown pills](https://github.com/localstack/localstack/blob/0b9f90a7221c3bfb76c4fd0c38531877ab968605/localstack/services/awslambda/invocation/version_manager.py#L95) and poison pills in the codebase.. 😄 👍